### PR TITLE
DROP SHARD

### PR DESCRIPTION
**Pull request by e-dard**
Addresses #5935.

The DROP SHARD command takes the ID of the shard and drops it
from both the meta store, and the data node. Shard data is deleted
on the node.

If the shard to be deleted is the last shard in the shard group,
then the shard group is marked as deleted.

Attempting to delete a shard that does not exist will _not_ return an
error.

- [x] CHANGELOG.md updated
- [x] Rebased/mergable
- [x] Tests pass
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

**Comment by corylanou** [comment]
+1

File: cluster/query_executor.go [hunk 1]
```diff
@@ -320,6 +320,16 @@ func (e *QueryExecutor) executeDropSeriesStatement(stmt *influxql.DropSeriesStat
 	return e.TSDBStore.DeleteSeries(database, stmt.Sources, stmt.Condition)
 }
 
+func (e *QueryExecutor) executeDropShardStatement(stmt *influxql.DropShardStatement) error {
+	// Remove the shard reference from the Meta Store.
+	if err := e.MetaClient.DropShard(stmt.ID); err != nil {
```
Comment (jwilder) [inline cluster/query_executor.go]: Should this come after the TSDBStore delete call?  If the TSDBStore delete fails for some reason, would you be able to retry and clean it up  again?

File: services/meta/client.go [hunk 2]
```diff
@@ -639,6 +639,21 @@ func (c *Client) ShardsByTimeRange(sources influxql.Sources, tmin, tmax time.Tim
...
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	data := c.cacheData.Clone()
+
+	if err := data.DropShard(id); err != nil {
+		return err
+	}
+
+	c.commit(data)
```
Comment (jwilder) [inline services/meta/client.go]: Error is ignored here.

**Comment by jwilder** [comment]
Two small things but :+1:

File: cluster/query_executor.go (hunk 1 above)
Comment (e-dard) [inline cluster/query_executor.go]: eurgh, `e.MetaClient.DropShard(stmt.ID)` doesn't need to return an error. I took the error out to ensure we can retry safely but didn't update the signature. Will update.

File: services/meta/client.go (hunk 2 above)
Comment (e-dard) [inline services/meta/client.go]: Seems to be ignored on every other call in the meta client too. Should they all be updated to check? @corylanou?

File: cluster/query_executor.go (hunk 1 above)
Comment (e-dard) [inline cluster/query_executor.go]: @jwilder :
> Should this come after the TSDBStore delete call? If the TSDBStore delete fails for some reason, would you be able to retry and clean it up again?

(sorry I think I deleted your comment by mistake). The meta store won't return an error if the shard does not exist, so it should be safe to call both these methods repeatedly.

**Pull request closed without merging**